qchem/tests/test_deriv_grad.py

```diff
@@ -0,0 +1,96 @@
+import numpy as np
+import pytest
+
+import pennylane as qml
+from pennylane import qchem
+
+import shutil
+
+
+non_zero_coeffs = np.array(
+    [
+        -0.7763135743293005,
+        -0.08534360840293387,
+        -0.08534360840293387,
+        0.2669341092545041,
+        0.26693410925450134,
+        -0.025233628744274508,
+        0.0072162443961340415,
+        -0.0072162443961340415,
+        -0.0072162443961340415,
+        0.0072162443961340415,
+        -0.030654287745411964,
+        -0.023438043349280003,
+        -0.023438043349280003,
+        -0.030654287745411964,
+        -0.02494407786332001,
+    ]
+)
+non_zero_ops = [
+    qml.Identity(wires=[0]),
+    qml.PauliZ(wires=[0]),
+    qml.PauliZ(wires=[1]),
+    qml.PauliZ(wires=[2]),
+    qml.PauliZ(wires=[3]),
+    qml.PauliZ(wires=[0]) @ qml.PauliZ(wires=[1]),
+    qml.PauliY(wires=[0]) @ qml.PauliX(wires=[1]) @ qml.PauliX(wires=[2]) @ qml.PauliY(wires=[3]),
+    qml.PauliY(wires=[0]) @ qml.PauliY(wires=[1]) @ qml.PauliX(wires=[2]) @ qml.PauliX(wires=[3]),
+    qml.PauliX(wires=[0]) @ qml.PauliX(wires=[1]) @ qml.PauliY(wires=[2]) @ qml.PauliY(wires=[3]),
+    qml.PauliX(wires=[0]) @ qml.PauliY(wires=[1]) @ qml.PauliY(wires=[2]) @ qml.PauliX(wires=[3]),
+    qml.PauliZ(wires=[0]) @ qml.PauliZ(wires=[2]),
+    qml.PauliZ(wires=[0]) @ qml.PauliZ(wires=[3]),
+    qml.PauliZ(wires=[1]) @ qml.PauliZ(wires=[2]),
+    qml.PauliZ(wires=[1]) @ qml.PauliZ(wires=[3]),
+    qml.PauliZ(wires=[2]) @ qml.PauliZ(wires=[3]),
+]
+
+
+x = np.array([0.0, 0.0, 0.35, 0.0, 0.0, -0.35])
+
+
+def H(x):
+    return qchem.molecular_hamiltonian(["H", "H"], x)[0]
+
+
+@pytest.mark.parametrize(
+    ("i", "coeffs", "ops"),
+    [(0, [], []), (2, non_zero_coeffs, non_zero_ops), (5, -non_zero_coeffs, non_zero_ops)],
+)
+def test_deriv(i, coeffs, ops):
+    r"""Tests the correctness of the nuclear derivative of the electronic Hamiltonian
+    computed by the ``'derivative'`` function. 
+    """
+
+    deriv = qchem.derivative(H, x, i)
+
+    shutil.rmtree("pyscf")
+
+    exp_obs = qml.vqe.Hamiltonian(coeffs, ops)
+
+    assert exp_obs.compare(deriv)
+
+
+def test_grad(tol):
```

Comment:
This is testing something more: it tests that the gradient of the energy is correct. I understand that the full gradient of the Hamiltonian will be a complicated object, but this test may fail if the gradient function works correctly but there are issues in other parts of the code, e.g., the circuit or `ExpvalCost`

Reply:
It also tests the integration with PL for computing expectation values of ``dh/dx``

Reply:
> I understand that the full gradient of the Hamiltonian will be a complicated object, but this test may fail if the gradient function works correctly but there are issues in other parts of the code, e.g., the circuit or ExpvalCost

Yes @ixfoduap, and this may be indeed a good thing. We experienced this type of problems with the ``grouping`` feature where we were able to compute expectation values but the energy was not correct. BTW, this is something pending, adding a unit test using the grouping feature.

Reply:
There is a difference between **unit tests** and **integration tests**. See [this](https://stackoverflow.com/questions/5357601/whats-the-difference-between-unit-tests-and-integration-tests) for a short explanation. Both are important. My point was that this is an integration test, but we also needed unit tests

Reply:
This test might look somehow forced, the problem is that as the ``gradient`` function is that simple testing the component of the gradients is redundant with the previous test testing the ``derivative`` function.